### ISSUE TITLE
Bug: [no-unsafe-enum-comparison] should only report on unsafe comparison to a literal

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8557]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.3.3&fileType=.tsx&code=KYOwrgtgBA0iD2B3EBlALgQzWAzsHUA3gFBRQDiA8pQCJQC8UA5FbUwDSlQBCAgnYyZ8aHYgF9ixAGZgQAYzQBLeCCgALDCAAmAG2DosuABQ5M2HAC4opgE6KQAcwCUVgG7xFWol0VSoJs1wGekY4JFRAvBwAOlYaJ28yMgB6ZKgteCgHeHgvNDV7By4xKGAdPChff1NDAhDQhGQDc3xo4QSSJKhU9MyAIww8gsdi0vLgRKSe2RtgOXgHewAvYC8a8wBuXqhNeHzgGyh8wtKQJVmdAE9i8SA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZonpNYAW1pkOY4nyrIOTdFETRoHaJHBgAviB1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
enum KnownStatuses {
  GOOD = 'GOOD',
  BAD = 'BAD',
}

function handleStatus(status: string): void {
  if (status === KnownStatuses.GOOD) {
    // do good thing
  } else if (status === KnownStatuses.BAD) {
    // do bad thing
  } else {
    // unrecoginzed status; do another thing entirely
  }
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unsafe-enum-comparison": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No errors. 

### Actual Result

Errors on both enum comparisons.

### Additional Info

I've just upgraded to typescript-eslint 7 in a new codebase, and I'm coming across dozens of examples of this sort of code that comes from handling network requests. I don't think there is anything wrong with the example code. This happened in the previous codebase I worked in as well when I upgraded the linter there.

I do not believe there is anything wrong with the example code, and I do not believe that it falls under the purpose of the rule as stated in the docs page.

I did find this closed issue (https://github.com/typescript-eslint/typescript-eslint/issues/7219) which described this pattern as a "niche" usage.... I'm not sure I agree with that assessment granted I've seen it many times in the wild. Furthermore, the reason that I'm opening this, rather than deferring to the closed issue, is that I'd like to make the case that a better solution would be to only compare against unsafe string _literals_ rather than using the type checker to compare against unsafe strings in general.

so 
```typescript
// still error
declare status: KnownStatuses;
if ('GOOD' === status) ...

// not error
declare status: string
if (status === KnownStatuses.GOOD) ...
```

**Comment by bradzacher** [comment]
My question here would be "why are you using an enum to check against plain strings?"

From your example, at least, it mostly looks like you're using an enum as a convenience way to create a namespaced string as opposed to some enumerated value. Which for sure - is a use of an enum - it's not a *super* common one.

The thing here is that this is an explicitly unsafe operation.
There's absolutely no guarantee that the string has any overlap with the enum.
From the types it seems equally likely that your code will always fall into the "else" branch (which from my experience is usually a `throw new Error("unexpected state")`.

Additionally this is unsafe because it's relying on an implementation detail - that the enum value is a string and that it has the "correct" value. There's a lot of space for things to break because there's nothing to enforce the two values are aligned - which is the definition of unsafe code!

To me at least; given that this is unsafe - it seems correct to report on this pattern? It seems correct to have a user disable the rule and say "I am purposely doing something unsafe here".

---

> I did find this closed issue (https://github.com/typescript-eslint/typescript-eslint/issues/7219) which described this pattern as a "niche" usage

Note that I wasn't describing the pattern of comparing an enum to a string as niche - I was describing wanting to ignore the check if and only if the LHS of the comparison is a non-enum and the RHS is an enum. That is a hyper-targeted opt-out and is thus niche.

**Comment by kirkwaiblinger** [comment]
I hear you. 

> I was describing wanting to ignore the check if and only if the LHS of the comparison is a non-enum and the RHS is an enum. That is a hyper-targeted opt-out and is thus niche.

Agreed that that specific opt-out strategy would probably not be justifiable for a library serving a general-purpose audience, as typescript-eslint does.

> "why are you using an enum to check against plain strings?"

I mean - you answered this exactly correctly, in a sense... to compare against a namespaced object. I think it seems like a pretty naturally occurring phenomenon to have something going on like this

```typescript
enum JobStatus {
  GOOD = 'GOOD',
  BAD = 'BAD',
}

// used for sending out data that we construct
declare function sendNetworkRequest(options: { status: JobStatus });

// used for parsing data that we receive
function parseNetworkResponse(response: { status: string, /* other data */ }) {
    if (response.status === JobStatus.GOOD) {
        // etc
    }
    // etc
}
```
where the parts of the application that _we control_ use the enum strictly (i.e. the creation and sending of a network request), and the parts we don't control, that are received from the outside world, are not known to be members of the enum (maybe they're even typed as `unknown`), but we expect them to be, and so we'll check at runtime. 

Is that the _best_ way to parse the unsafe data in a type-safe way? No, definitely not, but I don't think it's an obscure or terribly suspect pattern either for people who aren't keeping up with the latest trends in TypeScript type-safe validation libraries (like zod and such). I guess I just think that being in the recommended config is a stretch IMO.

More importantly, though, I think there's an opportunity here for us to improve the docs, at a minimum, and I'd like your input on this. The way I see it, there's two distinct purposes for the rule,

1. Make a literal-to-enum comparison resilient to refactoring.
   When refactoring  
   ```typescript
   enum Status {
       GoodStatus = 'true';
       BadStatus = 'false';
   }
   ```
   to 
   ```typescript
   enum Status {
       GoodStatus = 'false';
       BadStatus = 'true';
   }
   ```
   we have 
   ```typescript
   // Good, doesn't break behavior
   if (status === Status.GoodStatus)
   
   // woopsies, literal now corresponds to different enum value
   if (status === 'true')
   ```
   (as you've pointed out, this is _only_ a refactoring issue, since TS treats all enums as union types now, and prohibits comparison of types with no overlap)
1. Prohibit comparing a _wider_ type to an enum, since it may be considered an inappropriate use of an enum. (This is the main example we've been discussing so far.)

The [doc page for no-unsafe-enum-comparison](https://typescript-eslint.io/rules/no-unsafe-enum-comparison/) addresses purpose 1, but it does not address or give examples of purpose 2 whatsoever. To me, it's quite concerning that the only thing mentioned in the docs is the subset of the rule that you've described as verging on "quite useless" 😆. 

So, would you be open to converting this to a ticket to improve the docs to address Purpose 2?

As is, my thought process as a user went
1. Upgrade eslint major version
1. Get ~50 no-unsafe-enum-comparison errors
1. Look at docs. Not one of them looks like an example in the docs.
1. Think for 10 seconds
1. Still don't understand
1. Disable rule completely
1. Rule is effectively quite useless :)

**Comment by kirkwaiblinger** [comment]
actually, serious question, how would you actually recommend fixing the example i've brought up to comply with the no-unsafe-enum-comparison? Thinking about how i would write a doc page with "correct" vs "incorrect" code.

**Comment by kirkwaiblinger** [comment]
@Josh-Cena 

Hmmm. I'm not immediately convinced that trading non-opaque enum values for magic strings is an improvement. I'll have to think about it further, though.

Regardless, still think the docs are deficient.

**Comment by ziofat** [comment]
How about add a configuration for this rule to support the second purpose?

**Comment by adam-marshall** [comment]
RE:

> it mostly looks like you're using an enum as a convenience way to create a namespaced string as opposed to some enumerated value. Which for sure - is a use of an enum - it's not a super common one.

It is widespread in the codebase I work in.

Some examples:

- StatusCodes https://www.npmjs.com/package/http-status-codes is a number-based enum. Example usage
`[StatusCodes.NOT_FOUND, StatusCodes.FORBIDDEN].includes(httpErrorResponse.status)`
`httpErrorResponse.status === StatusCodes.NOT_FOUND` 

- Many examples where string-based unique identifiers from the server are represented as enum values for easier reference in the front-end codebase. Enums provide a useful shorthand for these.

**Comment by Josh-Cena** [comment]
> Many examples where string-based unique identifiers from the server are represented as enum values for easier reference in the front-end codebase. Enums provide a useful shorthand for these.

This is the use case I disagree with. Either your types say the values are `string`, and you have to handle unrecognized identifiers everywhere, or you do validation once at the input boundary (in which case you can temporarily disable this rule), and the resulting type is always an `enum` or a string literal union.

**Comment by kirkwaiblinger** [comment]
> This is the use case I disagree with. 

Personally, my opinion on the non-opaque enums is neutral for now. However, my opinion on a plan of action is that even if I agreed that using non-opaque enums is Very Objectionable, we'd need to 

either
- reluctantly support that use case
- OR Document conspicuously in the "when not to use" section that this rule is basically senseless if you use non-opaque enums

since I believe that this use case is common enough to warrant acknowledgement (and I also note that this issue has received an unusual amount of non-maintainer activity).

AND

- update the docs to give examples of comparison to both literals and non-literals, since the current docs just don't explain what the rule actually does at all.

**Comment by bradzacher** [comment]
My 2c is the same as Josh.
The reason so many people dislike enums is because they treat them as an "object container of known values" rather than a "set of opaque labels".

If you're talking API boundary conversion then as Josh mentioned you want this:
```ts
enum Foo { A = "A", ... }

/* eslint-disable @typescript-eslint/no-unsafe- enum-comparison */
function toFoo(value: string): Foo {
  switch (value) {
    case Foo.A:
      return Foo.A;
    //...
    default:
      throw new Error("unexpected value");
  }
}
/* eslint-enable @typescript-eslint/no-unsafe- enum-comparison */
```

And from that point forward the string is strictly typed as the opaque enum value and the rule won't ever error again.

We should definitely document this as a known edge case where a disable is recommended. It's a rare, valid case of intentionally piercing the opacity.
And that is in a nutshell what the rule enforces - "enums are opaque - don't use their values directly".

**Comment by Josh-Cena** [comment]
```js
case Foo.A:
      return Foo.A;
```

I suggested `case "A": return Foo.A`, this means even if you change it to `A = "a"`, nothing else needs changing.

**Comment by kirkwaiblinger** [comment]
Opening to PRs for docs changes only 👍
